sdk/go/common/workspace/workspace.go

```diff
@@ -139,8 +139,30 @@ func (pw *projectWorkspace) Save() error {
 	if err != nil {
 		return err
 	}
+	return atomicWriteFile(settingsFile, b)
+}
+
+// atomicWriteFile provides a rename based atomic write through a temporary file.
+func atomicWriteFile(path string, b []byte) error {
+	tmp, err := ioutil.TempFile("", filepath.Base(path))
+	if err != nil {
+		return errors.Wrapf(err, "failed to create temporary file %s", path)
+	}
+	defer func() { contract.Ignore(os.Remove(tmp.Name())) }()
 
-	return ioutil.WriteFile(settingsFile, b, 0600)
+	if err = tmp.Chmod(0600); err != nil {
+		return errors.Wrap(err, "failed to set temporary file permission")
+	}
+	if _, err = tmp.Write(b); err != nil {
+		return errors.Wrap(err, "failed to write to temporary file")
+	}
+	if err = tmp.Sync(); err != nil {
+		return err
+	}
+	if err = tmp.Close(); err != nil {
+		return err
+	}
+	return os.Rename(tmp.Name(), path)
```

Comment:
`os.Rename` works only if you use the same filesystem. in the OS'es where tmpfs is used and you trying to move somewhere else - this should fail.  
https://stackoverflow.com/questions/42392600/oserror-errno-18-invalid-cross-device-link
https://github.com/jupyterlab/jupyterlab/issues/5781

In the example for fedora:
```
 df -T /tmp/go-build* | awk '{print $2}' | tail -n1
tmpfs
df -T /usr/local/bin/stuff | awk '{print $2}' | tail -n1                                                                                                                                                                                                                                         
btrfs
```
In this scenario, this should fail so you might need to do one more sync from one file to other or check underlays fs types (atomic file writes are not so easy :/ ) 

p.s sorry if not the place and not the case, just came across this random PR on the internet :)